nncf/tensorflow/pruning/export_helpers.py

```diff
@@ -227,9 +228,46 @@ def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph):
         node.data['output_mask'] = input_masks[0]
 
 
+@TF_PRUNING_OPERATOR_METATYPES.register('reshape')
+class TFReshapeOps(DefaultMetaOp):
+    additional_types = ['Reshape']
+
+    @classmethod
+    def accept_pruned_input(cls, node: NNCFNode):
+        def _filtered_counter(shape):
+            filtered = [dim for dim in shape if dim not in [None, 1]]
+            return Counter(filtered)
+
+        input_shape = node.layer_attributes.input_shape
+        output_shape = node.layer_attributes.output_shape
+        # Check if this reshape is just squeeze / expand or transpose
+        return _filtered_counter(input_shape) == _filtered_counter(output_shape)
+
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph):
+        if cls.accept_pruned_input(node):
+            identity_mask_propagation(node, graph)
+        else:
+            node.data['output_mask'] = None
+
+
+@TF_PRUNING_OPERATOR_METATYPES.register('linear')
+class TFLinear(DefaultMetaOp):
+    additional_types = ['Dense']
+
+    @classmethod
+    def accept_pruned_input(cls, node: NNCFNode):
+        return True
+
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph):
+        output_mask = node.data.get('output_mask', None)
+        node.data['output_mask'] = output_mask
+
+
 @TF_PRUNING_OPERATOR_METATYPES.register('stop_propagation_ops')
 class TFStopMaskForwardOps(DefaultMetaOp):
-    additional_types = ['Dense', 'MatMul']
+    additional_types = ['MatMul']
```

Comment:
Is it potentially break pruning of the models with Dense layers in the classifier producing vector with constant number of classes. I think we still can't prune such layers. Will it be handled in this PR?

Reply:
Yes, I'll add tests to cover such cases. I think it should be the same logic as for the first convolution to prevent input image reduce but now it's about output classes